File: packages/flutter/ios/Classes/WaveUnlockPlugin.swift
import Flutter
import WaveUnlock  // the Swift core (pod / SPM dependency)

// Flutter plugin: runs the Swift UnlockEngine and streams each state over an
// EventChannel; startUnlock is triggered from a MethodChannel.
public class WaveUnlockPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    private var task: Task<Void, Never>?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = WaveUnlockPlugin()
        let methods = FlutterMethodChannel(name: "wave_unlock/control", binaryMessenger: registrar.messenger())
        let events = FlutterEventChannel(name: "wave_unlock/states", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methods)
        events.setStreamHandler(instance)
    }

    public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        sink = eventSink; return nil
    }
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        task?.cancel(); sink = nil; return nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "startUnlock", let a = call.arguments as? [String: Any],
              let pub = a["publishableKey"] as? String, let user = a["userNumber"] as? String
        else { result(FlutterMethodNotImplemented); return }

        let url = (a["gatewayUrl"] as? String).flatMap(URL.init) ?? WaveConfig.defaultGatewayURL
        Wave.configure(WaveConfig(publishableKey: pub, userNumber: user, gatewayURL: url))
        task?.cancel()
        task = Task { [weak self] in
            for await state in Wave.unlock() { self?.sink?(Self.encode(state)) }
        }
        result(nil)
    }

    private static func encode(_ s: UnlockState) -> [String: Any] {
        switch s {
        case .idle: return ["kind": "idle"]
        case .scanning: return ["kind": "scanning"]
        case .readerFound(let r): return ["kind": "readerFound", "rssi": r]
        case .tooFar(let r): return ["kind": "tooFar", "rssi": r]
        case .writing: return ["kind": "writing"]
        case .awaitingConfirmation: return ["kind": "awaitingConfirmation"]
        case .granted(let reason): return ["kind": "granted", "reason": reason as Any]
        case .denied(let reason): return ["kind": "denied", "reason": reason]
        case .timedOut: return ["kind": "timedOut"]
        case .failed(let e): return ["kind": "failed", "error": "\(e)"]
        }
    }
}

File: packages/flutter/android/src/main/kotlin/com/wave/unlock/flutter/WaveUnlockPlugin.kt
package com.wave.unlock.flutter

import android.content.Context
import com.wave.unlock.HttpGateway
import com.wave.unlock.UnlockEngine
import com.wave.unlock.UnlockState
import com.wave.unlock.WaveConfig
import com.wave.unlock.WaveProtocol
import com.wave.unlock.android.AndroidBleTransport
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

// Flutter plugin: runs the Kotlin UnlockEngine and streams each state over an
// EventChannel; startUnlock arrives on a MethodChannel.
class WaveUnlockPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var context: Context
    private var methods: MethodChannel? = null
    private var events: EventChannel? = null
    private var sink: EventChannel.EventSink? = null
    private val scope = CoroutineScope(Dispatchers.Main + Job())
    private var job: Job? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methods = MethodChannel(binding.binaryMessenger, "wave_unlock/control").apply { setMethodCallHandler(this@WaveUnlockPlugin) }
        events = EventChannel(binding.binaryMessenger, "wave_unlock/states").apply { setStreamHandler(this@WaveUnlockPlugin) }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methods?.setMethodCallHandler(null); events?.setStreamHandler(null); job?.cancel()
    }

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) { this.sink = sink }
    override fun onCancel(arguments: Any?) { job?.cancel(); sink = null }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "startUnlock") { result.notImplemented(); return }
        val cfg = WaveConfig(
            publishableKey = call.argument("publishableKey") ?: "",
            userNumber = call.argument("userNumber") ?: "",
            gatewayUrl = call.argument("gatewayUrl") ?: "https://app.wavepassport.com/api",
        )
        val engine = UnlockEngine(AndroidBleTransport(context), gateway = HttpGateway(cfg))
        job?.cancel()
        job = scope.launch {
            engine.unlock(WaveProtocol.payload(cfg.userNumber)).collect { sink?.success(encode(it)) }
        }
        result.success(null)
    }

    private fun encode(s: UnlockState): Map<String, Any?> {
        val m = mutableMapOf<String, Any?>("kind" to s.label)
        when (s) {
            is UnlockState.ReaderFound -> m["rssi"] = s.rssi
            is UnlockState.TooFar -> m["rssi"] = s.rssi
            is UnlockState.Granted -> m["reason"] = s.reason
            is UnlockState.Denied -> m["reason"] = s.reason
            is UnlockState.Failed -> m["error"] = s.error.name
            else -> {}
        }
        return m
    }
}
